aten/src/ATen/core/op_registration/op_registration_test.cpp

```diff
@@ -219,4 +220,265 @@ TEST(OperatorRegistrationTest, givenOpWithMultipleKernels_whenKernelsHaveSameDis
   }, "Tried to register multiple kernels with same dispatch key");
 }
 
+
+/**
+ * This is used to check that a given type works correctly when passed as input
+ * to or as output from a kernel.
+ *
+ * Call ArgTypeTestKernel<Input, Output>::test(input, inputExpectation, output, outputExpectation, schema)
+ * to test that a kernel with `Input` as input type and `Output` as output types,
+ * when called with `input` fulfills `inputExpectation` inside the kernel, then
+ * returns `output` and the returned value fulfills `outputExpectation`.
+ *
+ * Optionally, you can specify the argument list part of a function schema
+ * (e.g. "(Tensor a) -> Tensor") as an additional argument to use when
+ * registering the kernel.
+ */
+template<class InputType, class OutputType = InputType>
+struct ArgTypeTestKernel final : OperatorKernel {
+  explicit ArgTypeTestKernel(InputType input, std::function<void(const InputType&)> inputExpectation, OutputType output)
+  : input_(std::move(input)), inputExpectation_(std::move(inputExpectation)), output_(std::move(output)) {}
+
+  OutputType operator()(InputType input) const {
+    inputExpectation_(std::move(input));
+    return output_;
+  }
+
+  static void test(InputType input, std::function<void(const InputType&)> inputExpectation, OutputType output, std::function<void(const IValue&)> outputExpectation, const std::string& schema = "") {
+    auto registry = c10::RegisterOperators().op("_test::my_op" + schema, kernel<ArgTypeTestKernel>(input, std::move(inputExpectation), std::move(output)));
+    auto op = Dispatcher::singleton().findSchema("_test::my_op", "");
+    ASSERT_TRUE(op.has_value()); // assert schema is registered
+    auto actualOutput = callOp(*op, std::move(input));
+    EXPECT_EQ(1, actualOutput.size());
+    outputExpectation(actualOutput[0]);
+  }
+
+private:
+  InputType input_;
+  std::function<void(const InputType&)> inputExpectation_;
+  OutputType output_;
+  std::string schema_;
+};
+
+TEST(OperatorRegistrationTest, testAvailableArgTypes) {
+  // primitive types
+  ArgTypeTestKernel<double>::test(
+    1.5, [] (const double& v) {EXPECT_EQ(1.5, v);},
+    2.5, [] (const IValue& v) {EXPECT_EQ(2.5, v.toDouble());});
```

Comment:
These tests seem pretty repetitive. Shouldn't we able to do this in even fewer lines?

Reply:
(The reason I'm asking about this, is that tests are code too, and have to be maintained as well, and so big, wordy tests that are repetitive mean it will take you longer to do changes that change API, as you have to now fix all of the tests too.)

Reply:
It's not just the individual test cases, but also the fact that you have blasted out tests for permutations which aren't implemented separately (they're just different template instantiations).

Reply:
it's a trade off between conciseness and readability. I could make the test cases more concise by using type parametrized tests, but I think that would render them less readable and if something fails it's harder to figure out why and where it fails.